code/modules/mining/laborcamp/laborshuttle.dm

```diff
@@ -6,10 +6,180 @@
 	possible_destinations = "laborcamp_away"
 	circuit = /obj/item/circuitboard/labor_shuttle/one_way
-	req_access = list( )
+
+/obj/machinery/computer/shuttle/labor/one_way/allowed(mob/M)
+	. = ..()
+	if(.)
+		return TRUE
+
+	for(var/obj/item/card/id/prisoner/prisoner_id in M)
+		if(!prisoner_id.goal)
+			continue //no goal? no shuttle
+
+		if(prisoner_id.mining_points >= prisoner_id.goal)
+			return TRUE //completed goal? call it
+
+	return FALSE //if we didn't match above, no shuttle call
+
+/**********************Prisoners' Console**************************/
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console
+	name = "prisoner shuttle controller"
+	desc = "A console used by released prisoners to move the shuttle."
+	icon = 'icons/obj/machines/mining_machines.dmi'
+	icon_state = "console"
+	density = FALSE
+	var/machinedir = SOUTH
+	var/obj/item/card/id/prisoner/inserted_id
+	var/obj/machinery/door/airlock/release_door
+	var/door_tag = "prisonshuttle"
+	var/obj/item/radio/intercom/announcer
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/Initialize()
+	. = ..()
+	announcer = new /obj/item/radio/intercom(null)
+	announcer.config(list("Security" = 0))
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/Destroy()
+	. = ..()
+	QDEL_NULL(announcer)
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/card/id/prisoner))
+		if(!inserted_id)
+			if(!user.unEquip(I))
+				return
+			I.forceMove(src)
+			inserted_id = I
+			to_chat(user, "<span class='notice'>You insert [I].</span>")
+			SStgui.update_uis(src)
+			return
+		else
+			to_chat(user, "<span class='notice'>There's an ID inserted already.</span>")
+		return
+	return ..()
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/attack_hand(mob/user)
+	ui_interact(user)
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/attack_ghost(mob/user)
+	attack_hand(user)
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "PrisonerShuttleConsole", name, 315, 150, master_ui, state)
+		ui.open()
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/ui_data(mob/user)
+	var/list/data = list()
+	var/can_go_home = FALSE
+
+	data["emagged"] = emagged
+	data["id_inserted"] = inserted_id != null
+	if(inserted_id)
+		data["id_name"] = inserted_id.registered_name
+		data["id_points"] = inserted_id.mining_points
+		data["id_goal"] = inserted_id.goal
+
+	if(check_auth())
+		can_go_home = TRUE
+
+	data["can_go_home"] = can_go_home
+
+	return data
+
+/obj/machinery/mineral/labor_prisoner_shuttle_console/ui_act(action, params)
+	if(..())
+		return
+
+	switch(action)
+		if("handle_id")
+			if(inserted_id)
+				if(!usr.put_in_hands(inserted_id))
```

Comment:
you should be able to access the `user` from the tgui datum

Reply:
It is easier to use the `usr` variable as it has repeated use throught the proc and is concise, most examples of other TGUI interfaces use `usr`. I can see no benefit from checking the tgui datum's `user` variable each time for this.

Reply:
Querying coder chat about it now

Reply:
i was told it was preferred to use user, changing now